src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +124,143 @@ def __init__(
 
+    def _mark_commit_ready(self):
+        """Removes completed futures at the beginning of
+        ongoing_billing_outcomes and adds them to ready_to_commit.
+
+        If a future has errored, log an error but add to ready_to_commit
+        anyways.
+        """
+        while self._ongoing_billing_outcomes:
+            produce_billing, metrics_msg = self._ongoing_billing_outcomes[0]
+            if not produce_billing.done():
+                break
+
+            if produce_billing.exception():
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=produce_billing.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
+
+            self._ongoing_billing_outcomes.popleft()
+            self._ready_to_commit[metrics_msg.partition] = Position(
+                metrics_msg.next_offset, datetime.now()
+            )
+            self._messages_ready_since_last_commit += 1
+
+    def _should_commit(self) -> bool:
+        if not self._ready_to_commit:
+            return False
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        if self._last_commit_time + self._max_batch_time <= datetime.now():
+            return True
+        return False
+
+    def _bulk_commit(self) -> None:
+        """Commits and clears the ready to commit queue."""
+        if self._messages_ready_since_last_commit == 0:
+            return
+        self._commit(self._ready_to_commit)
+        self._clear_ready_queue()
+
+    def _clear_ready_queue(self) -> int:
+        """Clears the ready to commit queue and returns the amount of items dropped."""
+        amount_dropped = self._messages_ready_since_last_commit
+        self._ready_to_commit = {}
+        self._messages_ready_since_last_commit = 0
+        self._last_commit_time = datetime.now()
+        return amount_dropped
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+        self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            if deadline and deadline > time.time():
+                items_left = self._messages_ready_since_last_commit + len(
+                    self._ongoing_billing_outcomes
+                )
+                logger.warning(f"join timed out, items in the queue left: {items_left}")
+                break
+            self._mark_commit_ready()
+            self._bulk_commit()
+
     def terminate(self) -> None:
         self.close()
+        amount_dropped = len(self._ongoing_billing_outcomes)
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            ongoing_work.cancel()
+        amount_dropped += self._clear_ready_queue()
+        if amount_dropped > 0:
+            logger.warning(f"terminated, items dropped: {amount_dropped}")
```

Comment:
Should we make this an `error`?

Reply:
IMO, `error` should be used when there's an actual error and there's no error here. Further, if we look at `terminate`'s docs:

> Close the processing strategy immediately, abandoning any work in progress. No more messages should be accepted by the instance after this method has been called.

The caller should expect to drop items (i.e. "abandon any work in progress"). If the caller doesn't want to drop items, it should use `close` and `join` instead.

Reply:
Your call. I was thinking it might be nice if we got a sentry error for this.

Reply:
I agree. I set the consumer's logging level to `warning`, so that we don't get errors on intended operations but get messages on Sentry (you can filter these with `level:warning`). Done in [a37d70e](https://github.com/getsentry/sentry/pull/40030/commits/a37d70ea705707a04f3b6f807e682a5207d6ae43).

Reply:
I removed setting the log level to `WARNING` in [f497da2](https://github.com/getsentry/sentry/pull/40030/commits/f497da2ee71e713222db7ab021948af4a2c1dd08), since it was overriding the provided log level. The default value is lower and unless explicitly marking a higher one, messages will still get to Sentry (if one intentionally sets a higher limit, it's intended to not get warnings).